crates/circuit/src/symbol_expr.rs

```diff
@@ -0,0 +1,1528 @@
+                Value::Complex(c) => c.re,
+            }
+            None => 0.0,
+        }
+    }
+    pub fn imag(&self) -> f64 {
+        match self.eval(true) {
+            Some(v) => match v {
+                Value::Real(_) => 0.0,
+                Value::Complex(c) => c.im,
+            }
+            None => 0.0,
+        }
+    }
+    pub fn complex(&self) -> Complex64 {
+        match self.eval(true) {
+            Some(v) => match v {
+                Value::Real(_) => 0.0.into(),
+                Value::Complex(c) => c,
+            }
+            None => 0.0.into(),
+        }
+    }
+
+    pub fn symbols(&self) -> HashSet<String> {
+        match self {
+            SymbolExpr::Symbol(e) => HashSet::<String>::from([e.name.clone()]),
+            SymbolExpr::Value(_) => HashSet::<String>::new(),
+            SymbolExpr::Unary(e) => e.symbols(),
+            SymbolExpr::Binary(e) => e.symbols(),
+        }
+    }
+
+    pub fn has_symbol(&self, param: String) -> bool {
+        match self {
+            SymbolExpr::Symbol(e) => e.name == param,
+            SymbolExpr::Value(e) => false,
+            SymbolExpr::Unary(e) => e.has_symbol(param),
+            SymbolExpr::Binary(e) => e.has_symbol(param),
+        }
+    }
+
+    pub fn rcp(self) -> SymbolExpr {
+        match self {
+            SymbolExpr::Symbol(e) => SymbolExpr::Value( Value::Real(1.0)) / SymbolExpr::Symbol(e),
+            SymbolExpr::Value(e) => SymbolExpr::Value( Value::Real(1.0)) / SymbolExpr::Value(e),
+            SymbolExpr::Unary(e) => SymbolExpr::Value( Value::Real(1.0)) / SymbolExpr::Unary(e),
+            SymbolExpr::Binary(ref e) => match e.op {
+                BinaryOps::Div => SymbolExpr::Binary( Arc::new( Binary{ op: e.op.clone(), lhs: e.rhs.clone(), rhs: e.lhs.clone()}) ),
+                _ => SymbolExpr::Binary( Arc::new( Binary{ op: BinaryOps::Div, lhs: SymbolExpr::Value( Value::Real(1.0)), rhs: self.clone()}) ),
+            }
+        }
+    }
+
+    pub fn conjugate(&self) -> SymbolExpr {
+        match self {
+            SymbolExpr::Symbol(e) => SymbolExpr::Symbol(e.clone()),
+            SymbolExpr::Value(e) => match e {
+                Value::Complex(c) => SymbolExpr::Value( Value::Complex(c.conj())),
+                _ => SymbolExpr::Value( e.clone()),
+            },
+            SymbolExpr::Unary(e) => SymbolExpr::Unary( Arc::new( Unary{ op: e.op.clone(), expr: e.expr.conjugate()}) ),
+            SymbolExpr::Binary(e) => SymbolExpr::Binary( Arc::new( Binary{ op: e.op.clone(), lhs: e.lhs.conjugate(), rhs: e.rhs.conjugate()}) ),
+        }
+    }
+
+    pub fn is_complex(&self) -> bool {
+        match self {
+            SymbolExpr::Symbol(_) => false,
+            SymbolExpr::Value(e) => match e {
+                Value::Complex(_) => true,
+                _ => false,
+            },
+            SymbolExpr::Unary(e) => e.expr.is_complex(),
+            SymbolExpr::Binary(e) => e.lhs.is_complex() || e.rhs.is_complex(),

```

Comment:
Could this get us into trouble in expressions like `1j * 1j`, i.e. where LHS and RHS are complex but the expression is not? (Same question for `is_real`)